### ISSUE TITLE
Type alias with literal integer parameter incorrectly flagged as "Expected class but received   Literal[N]"

### ISSUE THREAD

**Issue by ds1sqe** [microsoft/pyright#11073]
## Environment

- **Pyright Version**: 1.1.407
- **Python Version**: 3.13.7
- **Operating System**: Linux 6.16.8-arch3-1
- **Project**: TypeBridge - TypeDB ORM for Python [[Link](https://github.com/ds1sqe/type-bridge)]

---

## Description

When using a type alias that takes an integer literal as a type parameter (following PEP 695 syntax), Pyright incorrectly reports a type error `"Expected class but received Literal[N]"` when the type alias is used as a field annotation in a class.

This issue prevents the use of integer literals as type parameters in generic type aliases, which is a valid pattern according to PEP 695.

---

## Minimal Reproduction

```python
from typing import Annotated

# Define a generic type alias with integer parameter
type Array[T, int] = Annotated[list[T], int]

class MyClass:
    arr: Array[str, 10]
    #                ^^
    # Error: Expected class but received "Literal[10]" [reportGeneralTypeIssues]
```

---

## TypeBridge Use Case

In the TypeBridge project, we use this pattern for cardinality constraints that map to TypeDB's `@card` annotations:

```python
from typing import Annotated

# Cardinality type aliases
type Min[MinRange, T] = Annotated[T, MinRange, "min"]
type Max[MaxRange, T] = Annotated[T, MaxRange, "max"]
type Range[MinRange, MaxRange, T] = Annotated[T, MinRange, MaxRange, "range"]

# Usage in entity definitions
class Person(Entity):
    tags: Min[2, Tag]              # Minimum 2 tags
    #         ^
    # Error: Expected class but received "Literal[2]"

    languages: Range[1, 5, Lang]   # Between 1 and 5 languages
    #                   ^
    # Error: Expected class but received "Literal[5]"
```

---

## Expected Behavior

The type alias should be accepted without errors, as integer literals are valid type parameters according to PEP 695. The following should all be valid:

```python
type Array[T, int] = Annotated[list[T], int]
arr: Array[str, 10]  # Should be OK - equivalent to Annotated[list[str], 10]

type Min[N, T] = Annotated[T, N, "min"]
field: Min[2, Tag]   # Should be OK - equivalent to Annotated[Tag, 2, "min"]
```

---

## Actual Behavior

Pyright reports the following error:

```
Expected class but received "Literal[10]" [reportGeneralTypeIssues]
```

The error occurs at the position of the integer literal in the type alias application.

---

## Runtime Behavior

The code works correctly at runtime:

```python
from typing import Annotated, get_type_hints, get_args, get_origin

type Min[N, T] = Annotated[T, N, "min"]

class Example:
    field: Min[2, str]

# Runtime introspection works correctly
hints = get_type_hints(Example, include_extras=True)
print(hints)  # {'field': typing.Annotated[str, 2, 'min']}

origin = get_origin(hints['field'])
args = get_args(hints['field'])
print(origin)  # <class 'str'>
print(args)    # (2, 'min')
```

---

## Impact

This issue affects:

1. **Type-safe cardinality constraints**: Cannot use integer literals to represent size constraints in type signatures
2. **Static type checking**: Forces use of `# type: ignore` comments or more complex workarounds
3. **Code clarity**: The pattern `Min[2, Tag]` is more readable than alternatives

---

### Option 1: Type ignore comments

```python
class Person(Entity):
    tags: Min[2, Tag]  # type: ignore[reportGeneralTypeIssues]
```

**Drawback**: Loses type checking benefits

### Option 2: Avoid integer literals as type parameters

This would require a completely different API design, losing the clean syntax.

### Option 3: Use string literals (not applicable)

Integer constraints cannot be represented as strings without losing semantic meaning.

---

## Related PEPs and Specifications

- **PEP 695**: Type Parameter Syntax - Introduces `type` statement for type aliases
- **PEP 593**: Flexible function and variable annotations - Defines `Annotated` type
- **PEP 646**: Variadic Generics - Discusses type parameters in general

According to PEP 695, type parameters can be any valid type expression, including literals.

---

### Why This Pattern is Useful

1. **Compile-time constraints**: Express size/cardinality constraints in the type system
2. **Better error messages**: Type checker can validate constraints statically
3. **Documentation**: Type signature documents the constraint
4. **Framework integration**: Maps cleanly to database schema constraints (TypeDB `@card` annotations)

### Similar Patterns in Other Libraries

This pattern is similar to:
- NumPy's shape annotations (when using typed arrays)
- SQLAlchemy's length constraints on String types
- Pydantic's `conint`, `constr` with min/max values

---

## Expected Fix

Pyright should recognize that when a type parameter is declared with a constraint (e.g., `int` in `type Array[T, int]`), literal values of that type are valid type arguments.

The type checker should:
1. Allow integer literals as type arguments when the type parameter accepts integers
2. Correctly resolve the type alias to its `Annotated` form
3. Preserve the literal value for runtime introspection

---

## Test Cases

```python
from typing import Annotated

# Test Case 1: Simple integer parameter
type Array[T, int] = Annotated[list[T], int]
arr1: Array[str, 10]        # Should be valid
arr2: Array[int, 100]       # Should be valid

# Test Case 2: Multiple integer parameters
type Range[Min, Max, T] = Annotated[T, Min, Max]
field1: Range[1, 10, str]   # Should be valid
field2: Range[0, 5, int]    # Should be valid

# Test Case 3: Mixed parameter types
type Tagged[T, str, int] = Annotated[T, str, int]
tagged1: Tagged[str, "label", 42]  # Should be valid

# Test Case 4: Nested usage
type Matrix[T, int, int] = Annotated[list[list[T]], int, int]
matrix: Matrix[float, 3, 3]  # Should be valid - 3x3 matrix
```

---

## Conclusion

This issue prevents the use of a clean, type-safe pattern for expressing compile-time constraints. A fix would enable better static type checking for frameworks that use integer literals as part of their type signature semantics.

The pattern is valid according to PEP 695 and works correctly at runtime, so Pyright should support it without reporting false positive errors.

---

## References

- PEP 695: https://peps.python.org/pep-0695/
- PEP 593: https://peps.python.org/pep-0593/
- Python Typing Documentation: https://docs.python.org/3/library/typing.html
---

**Comment by erictraut** [comment]
Pyright is working correctly here and is in compliance with PEP 695 and the Python typing spec.

Integer literal expressions are not allowed in a type expression. For details about what is allowed in a type expression, refer to [this section](https://typing.python.org/en/latest/spec/annotations.html#type-and-annotation-expressions) of the Python typing spec.

While `10` is not a valid type expression, `Literal[10]` is. Likewise, `Array[str, 10]` is not a valid type expression, but `Array[str, Literal[10]]` is valid.

BTW, I'm the author of PEP 695.

**Comment by ds1sqe** [comment]
@erictraut Thanks for your answer, And I misunderstand a "Literal Expression" with "Literal".

**Comment by erictraut** [comment]
Here are some additional details, in case they're helpful...

The statement `type Array[T, int] = Annotated[list[T], int]` does not "define a generic type alias with integer parameter". It defines a generic type alias with two type parameters named `T` and `int`. Neither of those type parameters have an explicit upper bound or value constraints, so they can take on any type. If you intend for the second type parameter to be constrained to `int` or subtypes thereof, you could define it as `type Array[T, S: int] = ...`. 

There is currently no way in the type system to constrain a type to an integer literal. There is a special form `LiteralString`, but there is no analogous `LiteralInt`. There has been some discussion about adding this, but to my knowledge no one has written a PEP yet.

I explored using `LiteralInt` for tensor shape analysis, but I ended up not pursuing the idea. If you're interested, here is a [document I wrote](https://docs.google.com/document/d/1UJKMW4QrRI9_KhK6vElDGOHGpoos0ZvrhxMX6GY8_a4/edit?tab=t.0#heading=h.ycactfd9sbhy).
